### ISSUE TITLE
Notes about new data entry screen

### ISSUE THREAD

**Issue by AdrienneRaniszewski** [ArctosDB/arctos#4114]
The feedback link takes me to issue #3691, which has been closed. So here are my notes (bugs, suggestions, gripes, etc.) 

1. I LOVE how customizable it is! The one thing I would like customized, but couldn’t figure out how to do is have our custom_ID (NK) show up somewhere prominent.

2. The show/carry settings only work for new records during the current session. If I’m starting fresh for the day I can either pick the collection, in which case nothing is filled out. Or I can start from a seed record, in which case everything is filled out. I’d like to the have the option back to start from a partially filled out record (the “carry” fields only)

3. I like that the mammal standard measurements can be added to the top of the attributes box. But I don’t like that you can’t quickly tab through those fields like we could before. It was a big time saver. I started to tab through but realized I was overwriting my entry.

4. Max error shouldn’t be a required field when entering coordinates.

5. When I use “pick locality,” “Locality ID,” or “No specific locality recorded” most of the fields do not automatically fill in like they did in the old screen. This is confusing and at first made me think I still needed to enter the info myself.

6. If I don’t select the coordinate units and instead just enter the coords in the appropriate fields, I am able to save without entering anything in datum, geo source, etc. I assume I will get yelled at during bulkloading for this, but it would be nice to get a warning at this stage instead (as the old form did)

7. I will probably get used to this, but the Place and Time box is hard on my eyes. I wish the sections had more visual separation. Even with getting rid of unused fields, it still feels cramped.

8. I also LOVE that all the extras can be entered right here on this screen! Man, that will speed things up so much. Also excited about the extras autoloading.

9. Very glad that I can set everything up and then our volunteers can just use my layout and customizations. Customizing every volunteer’s individual layout was a pain.

10. I read through the old issue, but don’t recall… is editing in this view something that’s in the works?

Overall, I think this new form is a nice improvement.

**Comment by dustymc** [comment]
>which has been closed

Yea, it was getting too big to be manageable, this is perfect.

>LOVE how customizable it is!

And I love how sustainable it is! (But that means it's probably always going to have some limitations/quirks....)

>have our custom_ID (NK) show up somewhere prominent.

Just carry other_id_type_1.

> start from a partially filled out record 

Enter what you want, change status to "@AdrienneRaniszewski 's left-handed shrew from barrow quad template don't touch this" and "seed" from it - you can have as many of those as you want/need. (Maybe it needs to be a bit more stable somehow, happy to talk about details, but I think the "build your own template" approach is pretty solid, whatever the details eventually turn out to be.) 

> mammal standard measurements 

I'm not sure I understand, elaborate please?

 >Max error shouldn’t be a required field when entering coordinates.

It should be coupled to distance and I can't make it twitchy, elaborate please.

>“pick locality,” “Locality ID,” or “No specific locality recorded” most of the fields do not automatically fill 

You can use pull/sync to fetch those data. I think that's generally less annoying than automagically pulling a bunch of values that then get ignored when it just uses the ID, but I'm not sure.


>If I don’t select the coordinate units and instead just enter the coords in the appropriate fields, I am able to save without entering anything in datum, geo source, etc. I assume I will get yelled at during bulkloading for this, but it would be nice to get a warning at this stage instead (as the old form did)

I think it will actually just not do anything, everything hinges on units. I'll see if I can make that check a little better.

>  I will probably get used to this, but the Place and Time box is hard on my eyes. I wish the sections had more visual separation. Even with getting rid of unused fields, it still feels cramped.

I'll play with that.

>is editing in this view something that’s in the works?

No, there's a new (very basic, feedback appreciated) dedicated form which does nothing else. Turning off {thing} and then using the same form to edit data which might have {thing} was a huge complexity, I think this approach is much simpler (=less likely to somehow result in mangled data), even if it means you see everything on the edit form.

>Overall, I think this new form is a nice improvement.

Thanks, so do I!

**Comment by Jegelewicz** [comment]
@dustymc when will make the switch to the new form? It would be super great if @AdrienneRaniszewski or someone else who has made good use of the new form could make a how to video about some of these bells and whistles so we can introduce it all before we just switch over.

**Comment by dustymc** [comment]
>switch

I have no compelling reason to make it go away, but I don't intend to update it either. Something a little more planned would be awesome.

**Comment by Jegelewicz** [comment]
It just seems that we have something that people like even better, but it's kinda hidden...

I do think it would be good to have documentation/instructions before we tell everyone to jump in. Maybe we should put this up for a discussion in AWG Issues.

**Comment by AdrienneRaniszewski** [comment]
have our custom_ID (NK) show up somewhere prominent.

Just carry other_id_type_1.

Okay, that’s what I’ve been doing.

start from a partially filled out record

Enter what you want, change status to ***@***.***<https://github.com/AdrienneRaniszewski> 's left-handed shrew from barrow quad template don't touch this" and "seed" from it - you can have as many of those as you want/need. (Maybe it needs to be a bit more stable somehow, happy to talk about details, but I think the "build your own template" approach is pretty solid, whatever the details eventually turn out to be.)

Interesting. But will the record save if I haven’t filled out all the “required” fields? I’ll give it a try.

mammal standard measurements

I'm not sure I understand, elaborate please?

I think it’s a function of how the mammal helper attributes are inputted now, but we used to be able to hit tab and move to the right to the next attribute field (for the standard mammal attributes only). It made entering those five measurements super fast. Related to that, neither form is quick to fill out. Have to move the mouse around a whole lot. Is it possible to have the attributes and parts react more like a table? “tab” across rows AND “return” down columns. We tend to enter attributes and parts by the column ( all part types, then all conditions, then all barcodes, etc), so being able to hit one key and move down to the next field would be amazing!

Max error shouldn’t be a required field when entering coordinates.

It should be coupled to distance and I can't make it twitchy, elaborate please.

Once I choose the coordinate unit, max error turns yellow. I can’t save the record unless I enter something there. It’s not uncommon that I don’t have that information. The old form let me leave that blank, which is preferable to making something up. Or what I’ve been doing is entering “0 m” and then editing it in SQL.

“pick locality,” “Locality ID,” or “No specific locality recorded” most of the fields do not automatically fill

You can use pull/sync to fetch those data. I think that's generally less annoying than automagically pulling a bunch of values that then get ignored when it just uses the ID, but I'm not sure.

Okay, great!

If I don’t select the coordinate units and instead just enter the coords in the appropriate fields, I am able to save without entering anything in datum, geo source, etc. I assume I will get yelled at during bulkloading for this, but it would be nice to get a warning at this stage instead (as the old form did)

I think it will actually just not do anything, everything hinges on units. I'll see if I can make that check a little better.

For now, I’ll just make sure everybody knows to select the units first.

Thanks!

**Comment by AdrienneRaniszewski** [comment]
sorry. that reply's a mess!

**Comment by AdrienneRaniszewski** [comment]
this is basically how I'd like a seed record filled out, but it won't save. maybe an option to "save as seed record" in addition to "Save as new Record". Set to not require any fields to be filled out. with status already set to "seed record - do not delete" or something.
<img width="2481" alt="Screen Shot 2021-11-17 at 10 35 55 AM" src="https://user-images.githubusercontent.com/23192428/142252967-aa0de6cc-f4a9-4752-859d-30bb02edcc63.png">

**Comment by dustymc** [comment]
>if I haven’t filled out all the “required” fields?

Worst case, you should be able to edit the record in the grid view - I don't think it cares about "required" at all, but let me know if that doesn't work for some reason. 

>maybe an option to "save as seed record"

Your timing is impeccable - I'll see what I can come up with.

>hit tab

Got it, I'll see what I can do.

>“tab” across rows AND “return” down columns

I suspect that's magic, but I'll see what I can come up with.

>Once I choose the coordinate unit

Got it.

**Comment by AdrienneRaniszewski** [comment]
just noticed that if I push the date and determiner from the top row of mammal helper measurements, it will only fill the first six attributes below, not the entire column. it'd be nice if it filled the entire column of attributes

**Comment by dustymc** [comment]
Coordinate dependencies should be a bit better and can be tested.

The "helper" thing is difficult; your network speed, local memory, and local CPU all seem to play a part, since changes call a whole bunch of other code which calls remote functions and sometimes calls more code and etc. - it's complicated. I've added a little pause, which may or may not do anything useful (or more annoying...) for your system. One possibility - and I'm not sure if it's realistic or not - would be to only do the processing on demand (eg click button, exit last field, whatever) after you've entered everything, rather than ever time there's a new value. I might also be able to make it a bit more tab-friendly by making/letting you type units instead of picking, but of course that opens up the possibility of an entirely new kind of problem. Let me know if any of that might be worth more exploration.

Parts tabbing might be a bit smoother now, but vertical is still beyond me. It's obviously more or less possible - Google Sheets does it in the same language. I'll keep an eye open to possibilities while waiting on the easy solution, Google's budget.... 

There's a new button:
<img width="333" alt="Screen Shot 2021-11-17 at 1 58 13 PM" src="https://user-images.githubusercontent.com/5720791/142289191-14aa6b06-f477-4a2d-9d51-2f2eaffa9d47.png">
<img width="484" alt="Screen Shot 2021-11-17 at 1 58 20 PM" src="https://user-images.githubusercontent.com/5720791/142289206-8688a244-26dc-44be-8d30-414de0d86829.png">

**Comment by AdrienneRaniszewski** [comment]
great! I'll be using the form some more today and will let you know how it goes

**Comment by dustymc** [comment]
>today

All of the new stuff is still in test.

**Comment by sharpphyl** [comment]
There seems to be a bug in the pull/sync event where there is a depth in the locality.  

Collecting event 11139041 has a depth of 60 feet.

![image](https://user-images.githubusercontent.com/15368365/144124580-8159e656-f111-4dfa-8557-51c6a634eb6c.png)

If I pull/sync Event 11139041 everything transfers except the depth.

![image](https://user-images.githubusercontent.com/15368365/144124401-088b4875-7156-47b8-8d0f-e594fc2cbd32.png)

If I clone from catalog record DMNS:Inv: 25539, both the event and locality, I still don't get the depth. 

If I manually add the locality ID 10913888, then I get the depth as well as the rest of the locality data.

![image](https://user-images.githubusercontent.com/15368365/144125117-8d5aad5d-d738-46f2-a968-2014f3ca4134.png)

Is the depth (as well as the rest of the locality data) supposed to be populated with the Collecting Event ID?

**Comment by dustymc** [comment]
I added depth to the pull for next release, but I'm not sure the situation is clear.

1. If a collecting_event_id or collecting_event_name is provided, then that event is just used. You can put "correct" data, incorrect data, clown emojis, nothing at all, WHATEVER in any other collecting event/locality/geography/related attributes field and it will do absolutely nothing.
2. Ditto locality, just with a few fewer fields.

It makes no difference where the identifier comes from - you can pick it, type it in, pull it from an existing record, and probably some other stuff. All that matters is whether there's an identifier in place when the record loads to the core DB (in which case it will be used)  or not (in which case some data are required).

If that's clear, yay everybody. If it's not, please let me know how I can make the popup better.

<img width="474" alt="Screen Shot 2021-12-07 at 2 21 11 PM" src="https://user-images.githubusercontent.com/5720791/145114997-3ea1fc52-6b73-4f0d-95f4-c7d4b8dd4e3a.png">

**Comment by dustymc** [comment]
>no compelling reason to make it go away

The old form won't control UTM Zone, and most existing data are garbage which suggests control is necessary. I'm not sure how compelling that is, but starting some "official" process to switch seems useful to me. @Jegelewicz

**Comment by Jegelewicz** [comment]
Adding to AWG agenda for tomorrow. I think we just need to announce it everywhere.

**Comment by sharpphyl** [comment]
> I added depth to the pull for next release, but I'm not sure the situation is clear.

The depth showed up on one pull/sync today, but I'll continue to monitor that.  Thanks for adding it.  

I get that the event ID will populate during the upload, but it's helpful to see that data in the entry form as well, especially as I have to approve then upload volunteer entries.  

I no longer see "depick event" and from your comment above, we have to remove the eventID to create a new event ID with changed data.  Correct?  You might say in the pop-up that changing data will not edit the event unless the collecting_event_id is depicked (or erased or depulled or whatever the new term would be).

**Comment by campmlc** [comment]
Yes, I like clear all and clear locality

**Comment by dustymc** [comment]
>clear all and clear locality

In next release (along with a clear event only option).

**Comment by sharpphyl** [comment]
I'm not sure about the "clear" functionality.  Usually, I want to keep all the event and locality data but make one small adjustment such as the depth or the date, so it will create a new event.  Depick used to just eliminate the event or locality ID and allow you to make changes.  Can "clear ID" do the same thing and not clear all the data.  Otherwise, there's no reason to pull from an event at all, but we encourage it for consistency.

**Comment by dustymc** [comment]
>"clear ID" 

Next release:
<img width="1075" alt="Screen Shot 2021-12-15 at 10 33 50 AM" src="https://user-images.githubusercontent.com/5720791/146245082-d667c1c8-1014-44d5-977f-3dca418fc3ba.png">

**Comment by sharpphyl** [comment]
Perfect.  Thanks.

**Comment by sharpphyl** [comment]
@dustymc  I'm developing a data entry checklist for our volunteers.   In the Locality section within Place and Time, could you reconsider the order of the fields.  This was the order in the previous UI, but there were three visually distinct sections for Specimen Event, Collecting Event and Locality and now they are merged into Place and Time.  The first field under Locality is Higher Geography and the next fields are Locality Name and Locality ID.  

<img width="623" alt="Screen Shot 2022-01-09 at 2 29 42 PM" src="https://user-images.githubusercontent.com/15368365/148702040-2ceb6f50-bab9-4ecc-b020-807412f8596a.png">

If you fill in a Locality Name or ID and pull/sync, it fills in the Higher Geography.  Shouldn't that field come after the two fields that can populate the entire Locality section?  That order would also make it identical in structure to the Collecting Event section.

<img width="644" alt="Screen Shot 2022-01-09 at 2 47 21 PM" src="https://user-images.githubusercontent.com/15368365/148702294-ad1ef88e-375d-4998-aace-6bf1a0829fd7.png">

BTW, creating and saving profiles is awesome!!  We'll have each volunteer create their own for each unique collection.

**Comment by dustymc** [comment]
The order is ~most to ~least specific. I think that's more or less how most human brains work, but it's easy enough to shuffle fields. I don't think geography somewhere in the middle makes much sense, if that's what you're suggesting.

Note that pull/sync exists just to give humans something to look at - all that's really necessary is providing collecting_event_name.

**Comment by sharpphyl** [comment]
I guess that's a "no."

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
I finally had a chance to enter some records through the new data entry screen and had a few issues with trying to enter part attributes:

I am unable to enter part attribute numeric order, I get an error message if I select it. If I hit ok, the value cell goes white and I can't enter anything. If I try to reselect numeric order, I get the error message again.
![Data Entry Capture 1](https://user-images.githubusercontent.com/47402809/148836271-c14ed506-9608-4c32-9800-6cb9a25afed0.PNG)


I am also having a problem with part attributes trying to sort of carry forward even though I don't have them set to carry forward. See, values carried forward:
![Data Entry Capture 4](https://user-images.githubusercontent.com/47402809/148837043-db82b3dc-e041-4619-bf5b-5b4560c4dd40.PNG)

Even though I don't have them set to carry forward:
![Data Entry Capture 3](https://user-images.githubusercontent.com/47402809/148837180-2a54b9c8-d1eb-44ec-8ed3-7eb6be9fdc7d.PNG)

And finally, it now won't let me enter less than the number of attributes I used in the previous record:
![Data Entry Capture 5](https://user-images.githubusercontent.com/47402809/148837326-994e4f26-91d0-4a29-9f31-1a54e96085f9.PNG)
![Data Entry Capture 6](https://user-images.githubusercontent.com/47402809/148837338-67ec8019-fce5-453c-a3c8-48693ca278c8.PNG)

**Comment by dustymc** [comment]
>guess that's a "no."

Certainly not intended that way.

>numeric order,

That attribute is not set up properly in test or prod, of course there's not documentation, and I can't find anything - what's supposed to be going on with that attribute, do you know where it came from, ????

Otherwise, the behavior should be better in test/next release.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
> I can't find anything

https://github.com/ArctosDB/arctos/issues/2330#issuecomment-554028123

**Comment by cjconroy** [comment]
some comments
1. if you forget the event determination date, it is noted, but the box is not highlighed as it was before. 
2. MVZ mammals is my default, but have some mammals from herp or bird accessions. When I put in the accession number, it says not found. I then have to select mammal then search again. Or I need to go to the search box. Since our accession numbers are not overlapping, seems like it would be OK to just have the correct accession come back if I just put in the number. 
3. when I select an agent under the identification and then click copy to all, that name does not pop up in the agent window. That person is very likely going to be the collector or preparator, so having it go in there would save time. That name also is not put into the standard mammal measurements row, but does show up in the table below it. That looks confusing.
4. After a few specimens with 3 parts, my current specimen has only 2 parts, but arctos won't proceed, telling me abort, must have all the components of part 3. Must I make one up?
5. what's the difference between a specimen and a preserved specimen? There is no link to a vocabulary. 
6. with the old form, if you left a value for an attribute, but not the name of an attribute, it would be OK upon save. But, now getting error that I need to clear that attribute. 
7. There is an option to go to your last record when starting over. Not clear that is an option now for this test screen. 
8. In regards to #6, I cleared the agent on the blank attribute. That, as usual, brings up a search box for an agent, which I usually click to close. On a mac, I just did command W and it closed the entire data entry screen.

**Comment by Jegelewicz** [comment]
different coordinate types are not supported? If I select deg. min. sec. there is no place to enter the values.

Also, once stuff is entered, what is this supposed to do?

![image](https://user-images.githubusercontent.com/5725767/152407340-dda42e7f-4ba9-436d-851f-fbe651f94dfe.png)

Selecting it gets me this

![image](https://user-images.githubusercontent.com/5725767/152407404-0d327326-905f-420e-a21d-546cb0e067d5.png)

**Comment by Jegelewicz** [comment]
Was working with someone to enter things last week and could find no way to hide the "simplified" parts section.

![image](https://user-images.githubusercontent.com/5725767/152407851-8b89e14a-9ac7-4d3c-8f02-060c894924e6.png)

My records all have parts with attributes so I am using the "extras Parts" section and the simplified one just means more scrolling.

**Comment by Jegelewicz** [comment]
When I select this form for data entry, I always end up down in the "extras parts" section why does it scroll down there first?

![image](https://user-images.githubusercontent.com/5725767/152408832-b373e2e6-2152-4cdc-bae9-9109d2b0d1a3.png)

**Comment by dustymc** [comment]
>attributes....

I rebuilt everything.

>if you forget the event determination date, it is noted, but the box is not highlighed as it was before.

I don't think highlighting is practical with this level of customization, but I can explore further if it's critical - that's probably new-issue worthy.

>    MVZ mammals is my default, but have some mammals from herp or bird accessions. When I put in the accession number, it says not found. I then have to select mammal then search again. Or I need to go to the search box. Since our accession numbers are not overlapping, seems like it would be OK to just have the correct accession come back if I just put in the number.

Recommend using the pick button - there's no way the processing (which sees everything) can happen with ambiguous accn numbers.

>    when I select an agent under the identification and then click copy to all, that name does not pop up in the agent window. That person is very likely going to be the collector or preparator, so having it go in there would save time. That name also is not put into the standard mammal measurements row, but does show up in the table below it. That looks confusing.

Collector_agent_1 is included in next release. The "helper" row is just that, it doesn't carry data, it just fills in the "data rows." (That is, I can add stuff there but it can't do anything.)

>    After a few specimens with 3 parts, my current specimen has only 2 parts, but arctos won't proceed, telling me abort, must have all the components of part 3. Must I make one up?

I think this has been fixed somewhere along the way, I need more information if you can still create this situation.

>    what's the difference between a specimen and a preserved specimen? There is no link to a vocabulary.

Lots of CT definition links in next release, that's always available from search/code tables.

>    with the old form, if you left a value for an attribute, but not the name of an attribute, it would be OK upon save. But, now getting error that I need to clear that attribute.

See above, rebuilt everything. 

>   There is an option to go to your last record when starting over. Not clear that is an option now for this test screen.
    In regards to 

Added. (You can also start from any record you can see.)

> https://github.com/ArctosDB/arctos/issues/6, I cleared the agent on the blank attribute. That, as usual, brings up a search box for an agent, which I usually click to close. On a mac, I just did command W and it closed the entire data entry screen.

I don't have any control over OS-level commands, but ``esc`` should close the agent pick on a mac (unless you or something in your software stack has changed that)

>why does it scroll 

Your browser is calculating the viewport before it gets everything fully rendered or similar, probably. It'll force-scroll to the top on window load in next release.
